packages/realm/src/app-services/BaseSubscriptionSet.ts

```diff
@@ -0,0 +1,245 @@
+        return DEFAULT_PROPERTY_DESCRIPTOR;
+      }
+    }
+  },
+  // Not defining `set()` here will make e.g. `mySubscriptions[0] = someValue` a no-op
+  // if strict mode (`"use strict"`) is used, or throw a TypeError if it is not used.
+};
+
+/**
+ * Class representing the common functionality for the {@link SubscriptionSet} and
+ * {@link MutableSubscriptionSet} classes.
+ *
+ * The {@link Subscription}s in a SubscriptionSet can be accessed as an array, e.g.
+ * `realm.subscriptions[0]`. This array is readonly – SubscriptionSets can only be
+ * modified inside a {@link SubscriptionSet.update} callback.
+ */
+export abstract class BaseSubscriptionSet {
+  // Enables index accessing when combined with the proxy handler's `get()` method
+  readonly [n: number]: Subscription;
+
+  /**@internal */
+  protected constructor(/**@internal */ protected internal: binding.SyncSubscriptionSet) {
+    Object.defineProperties(this, {
+      internal: {
+        enumerable: false,
+        configurable: false,
+        // `internal` needs to be writable due to `SubscriptionSet.updateSync`
+        // overwriting `this.internal` with the new committed set.
+        writable: true,
+      },
+    });
+    return new Proxy<BaseSubscriptionSet>(this, PROXY_HANDLER);
+  }
+
+  /**
+   * Determine whether the SubscriptionSet is empty.
+   *
+   * @returns `true` if there are no subscriptions in the set, otherwise `false`.
+   */
+  get isEmpty(): boolean {
+    return this.internal.size === 0;
+  }
+
+  /**
+   * Get the version of the SubscriptionSet.
+   *
+   * @returns The version of the SubscriptionSet. This is incremented every time an
+   *  {@link SubscriptionSet.update} is applied.
+   */
+  get version(): number {
+    return Number(this.internal.version);
+  }
+
+  /**
+   * Get the state of the SubscriptionSet.
+   *
+   * @returns The state of the SubscriptionSet.
+   */
+  get state(): SubscriptionsState {
+    const stateValue = this.internal.state.valueOf();
+    switch (stateValue) {
+      case 0: // Uncommitted
+      case 1: // Pending
+      case 2: // Bootstrapping
+      case 6: // AwaitingMark
+        return SubscriptionsState.Pending;
+      case 3:
+        return SubscriptionsState.Complete;
+      case 4:
+        return SubscriptionsState.Error;
+      case 5:
+        return SubscriptionsState.Superseded;
+      default:
+        throw new Error(`Unsupported SubscriptionsState value: ${stateValue}`);
+    }
+  }
+
+  /**
+   * Get the state of the SubscriptionSet.
+   *
+   * @returns If `state` is {@link SubscriptionsState.Error}, this will return a `string`
+   *  representing why the SubscriptionSet is in an error state. `null` is returned if there is no error.
+   */
+  get error(): string | null {
+    return this.state === SubscriptionsState.Error ? this.internal.errorStr : null;
+  }
+
+  /**
+   * Get the number of subscriptions in the set.
+   *
+   * @returns The number of subscriptions in the set.
+   * @internal
```

Comment:
It was internal before and not part of the [types](https://github.com/realm/realm-js/blob/d592f46c031ea62da97650a68f4ac4757d1dbd83/types/index.d.ts#L876). I believe the legacy code was using this.

But yes, to extend `ReadonlyArray` this will be public. I'm adding those methods in a separate PR 👍  (even w/o the array behavior, a `length` prop would prob be appreciated like you said)